content/en/blog/_posts/2022-11-21-devicemanager-ga.md/index.md

```diff
@@ -0,0 +1,89 @@
+     options to the `DeviceManager` in order to indicate if `PreStartContainer`,
+     `GetPreferredAllocation` or other future optional calls are supported and
+     can be called before making devices available to the container.
+  1. `GetPreferredAllocation` allows a device plugin to forward allocation
+     preferrence to the `DeviceManager` so it can incorporate this information
+     into its allocation decisions. The `DeviceManager` will call out to a
+     plugin at pod admission time asking for a preferred device allocation
+     of a given size from a list of available devices to make a more informed
+     decision. E.g. Specifying inter-device constraints to indicate preferrence
+     on best-connected set of devices when allocating devices to a container.
+  1. `PreStartContainer` is called before each container start if indicated by
+     device plugins during registration phase. It allows Device Plugins to run device
+     specific operations on the Devices requested. E.g. reconfiguring or
+     reprogramming FPGAs before the container starts running. 
+
+Pull Requests that introduced these changes are here:
+1. [Invoke preStart RPC call before container start, if desired by plugin](https://github.com/kubernetes/kubernetes/pull/58282)
+1. [Add GetPreferredAllocation() call to the v1beta1 device plugin API](https://github.com/kubernetes/kubernetes/pull/92665)
+
+With introduction of the above endpoints the interaction between Device Manager in
+kubelet and Device Manager can be shown as below:
+
+{{< figure src="deviceplugin-framework-overview.svg" alt="Device Plugin Framework Figure" class="diagram-large" caption="Device Plugin Framework Overview" link="https://kubernetes.io/blog/2022/11/21/devicemanager-graduates-to-GA/" >}}
+
+### Change in semantics of device plugin registration process 
+Device plugin code was refactored to separate 'plugin' package under the `devicemanager`
+package to lay the groundwork for introducing a `v1beta2` device plugin API. This would
+allow adding support in `devicemanager` to service multiple device plugin APIs at the
+same time.
+
+With this refactoring work, it is now mandatory for a device plugin to start serving its gRPC
+service before registering itself with kubelet. Previously, these two operations were asynchornous
+and device plugin could register itself before starting its gRPC server which is no longer the
+case. For more details, refer to [PR #109016](https://github.com/kubernetes/kubernetes/pull/109016) and [Issue #112395](https://github.com/kubernetes/kubernetes/issues/112395).
+
+### Dynamic resource allocation
+In Kubernetes 1.26, inspired by how [Persistent Volumes](/docs/concepts/storage/persistent-volumes) are handled in Kubernetes, [Dynamic Resource Allocation](/docs/concepts/scheduling-eviction/dynamic-resource-allocation.md) has been introduced to cater to devices that have more sophisticated resource requirements
```

Comment:
```suggestion
In Kubernetes 1.26, inspired by how [Persistent Volumes](/docs/concepts/storage/persistent-volumes)
are handled in Kubernetes, [Dynamic Resource Allocation](/docs/concepts/scheduling-eviction/dynamic-resource-allocation/)
has been introduced to cater to devices that have more sophisticated resource requirements
```

The hyperlink to `/docs/concepts/scheduling-eviction/dynamic-resource-allocation.md` is wrong; I fixed that as well as wrapping the text (friendlier to localization teams).

Reply:
Thanks, fixed.